core/sql/optimizer/RelExpr.cpp

```diff
@@ -9121,38 +9090,85 @@ NABoolean Scan::updateableIndex(IndexDesc *idx)
     CMPASSERT(updateCol->getOperatorType() == ITM_BASECOLUMN);
-
     for (CollIndex j = 0; j < indexKey.entries(); j++)
     {
       ItemExpr *keyCol = indexKey[j].getItemExpr();
-
-      ItemExpr *baseCol = ((IndexColumn*)keyCol)->getDefinition().getItemExpr();
+      ItemExpr *baseCol = 
+        ((IndexColumn*)keyCol)->getDefinition().getItemExpr();
       CMPASSERT(baseCol->getOperatorType() == ITM_BASECOLUMN);
-
-      // QSTUFF
-      if (getGroupAttr()->isEmbeddedUpdate()){
-        if (((BaseColumn*)updateCol)->getColNumber() ==
+      if (((BaseColumn*)updateCol)->getColNumber() ==
           ((BaseColumn*)baseCol)->getColNumber())
-          return FALSE;
-      }
-      // QSTUFF
+        return FALSE;
+    }
+  }
+  return TRUE;
+} // Scan::updateableIndex
+
+NABoolean Scan::requiresHalloweenForUpdateUsingIndexScan()
+{
+
+  // Returns TRUE if any non clustering index can be used for a scan 
+  // during an UPDATE and a key column of that index is in the SET
+  // clause of the update. If this method returns TRUE we will use a 
+  // sort node to prevent the "Halloween Update Problem".
+  //
 
-      if ((NOT(idx->isUniqueIndex() || idx->isClusteringIndex()))
-	     ||
-	  (CmpCommon::getDefault(UPDATE_CLUSTERING_OR_UNIQUE_INDEX_KEY) == DF_OFF))
+  // preds are in RangeSpec form
+  ValueIdSet preds = getSelectionPredicates();
+  const ValueIdList colUpdated = getTableDesc()->getColUpdated();
+  const LIST(IndexDesc *) & ixlist = getTableDesc()->getIndexes();
+
+  if ((colUpdated.entries() == 0) || (preds.entries() == 0) ||
+      (ixlist.entries() == 1) || // this is the clustering index
+      (CmpCommon::getDefault(UPDATE_CLUSTERING_OR_UNIQUE_INDEX_KEY) 
+       == DF_AGGRESSIVE)) // this setting means no Halloween protection
+    return FALSE;
+
+   if (CmpCommon::getDefault(RANGESPEC_TRANSFORMATION) == DF_ON)
+   {
+     ValueIdList selectionPredList(preds);
+     ItemExpr *inputItemExprTree = 
+       selectionPredList.rebuildExprTree(ITM_AND,FALSE,FALSE);
+     ItemExpr * resultOld = revertBackToOldTree(STMTHEAP, 
+                                                inputItemExprTree);
+     preds.clear();
+     resultOld->convertToValueIdSet(preds, NULL, ITM_AND);
+     doNotReplaceAnItemExpressionForLikePredicates(resultOld,preds,
+                                                   resultOld);
+   }
 
+  for (CollIndex indexNo = 0; indexNo < ixlist.entries(); indexNo++)
+  {
+    IndexDesc *idx = ixlist[indexNo];
+    if (idx->isClusteringIndex() || 
+        (idx->isUniqueIndex() && 
+        (CmpCommon::getDefault(UPDATE_CLUSTERING_OR_UNIQUE_INDEX_KEY) 
+         == DF_ON)))
+      continue ; // skip this idesc
+
+    const ValueIdList indexKey = idx->getIndexKey();
+    // Determine if the columns being updated are key columns. Each key
+    // column being updated must have an associated equality clause in
+    // the WHERE clause of the UPDATE for it to be used.
+    for (CollIndex i = 0; i < colUpdated.entries(); i++)
```

Comment:
Yes. Please see response above. We  should now accept any index.  Certain index/predicate combinations will require a blocking sort. We will make the choice to create that sort, even if that particular index is not chosen.